frontend/ui/data/keyboardlayouts/zh_ime.lua

```diff
@@ -0,0 +1,346 @@
+                if value2 == value then
+                    if fcompval( tbl[iMid-1] ) ~= value then
+                        return iMid
+                    else
+                        iEnd = iMid - 1
+                    end
+                else
+                    if fcompval( tbl[iMid+1] ) == value then
+                        return iMid + 1
+                    else
+                        iStart = iMid + 2
+                    end
+                end
+            end
+            return iMid
+        elseif ( reversed and value2 < value ) or ( not reversed and value2 > value ) then
+            iEnd = iMid - 1
+        else
+            iStart = iMid + 1
+        end
+    end
+end
+
+local function stringReplaceAt(str, pos, r)
+    return str:sub(1, pos-1) .. r .. str:sub(pos+1)
+end
+
+local _stack
+local IME = {
+    code_map = {},
+    key_map = nil, -- input key to code map
+    keys_string = "abcdefghijklmnopqrstuvwxyz",
+    iter_map = nil, -- next code when using wildcard
+    iter_map_last_key = nil,
+    show_candi_callback = function() end,
+    switch_char = "换字", -- default
+    separator = "分隔",  -- default
+    use_space_as_separator = true,
+    local_del = "",  -- default
+    W = nil -- default no wildcard
+}
+
+function IME:new(new_o)
+    local o = new_o or {}
+    setmetatable(o, self)
+    self.__index = self
+    o:init()
+    return o
+end
+
+function IME:init()
+    self:clear_stack()
+
+    self.sorted_codes = {}
+    for k,_ in pairs(self.code_map) do
+        table.insert(self.sorted_codes, k)
+    end
+    table.sort(self.sorted_codes)
+
+    if not self.key_map and self.keys_string then
+        self.key_map = {}
+        for i=0, #self.keys_string do
+            self.key_map[self.keys_string:sub(i,i)] = self.keys_string:sub(i, i)
+        end
+    end
+
+    if not self.iter_map and self.W then
+        self.iter_map = {}
+        local keys = util.splitToChars(self.keys_string)
+        for i=1, #keys-1 do
+            self.iter_map[keys[i]] = keys[i+1]
+        end
+        if #keys > 1 then
+            self.iter_map[keys[#keys]] = keys[1]
+        end
+    end
+end
+
+function IME:clear_stack()
+    _stack = { {code="", char="", index=1, candi={} } }
+    self.last_key = ""
+    self.last_index = 0
+    self.hint_char_count = 0
+end
+
+function IME:reset_status()
+    self.last_key = ""
+    self.last_index = 0
+end
+
+function IME:searchStartWith(code)
+    local result = binarysearch(self.sorted_codes, code, function(v) return string.sub(v or "", 1, #code) end)
+    if result then
+        local candi = self.code_map[self.sorted_codes[result]]
+        if candi then
+            logger.dbg("zh_kbd: got search result starting with", code, ":", candi)
+        end
+        if type(candi) == "string" then
+            return { candi }
+        end
+        return candi
+    end
+end
+
+function IME:getCandiFromMap(code)
+    local candi = self.code_map[code]
```

Comment:
You can use "Candidate" (at least in the method/function names, less critical for local variables), for readability.
(Unless this means candy/sweets :)

Reply:
Candi's sweet, I think I'm keeping it ;) Also it shows a lot in this file, trying to save some space:)